sycl/doc/design/PropertySets.md

```diff
@@ -0,0 +1,245 @@
+# SYCL binary property sets
+
+To communicate information about SYCL binaries to the SYCL runtime, the
+implementation produces sets of properties. The intention of this design
+document is to describe the structure of the property sets and define the
+representation and meaning of pre-defined property set names.
+
+## Property sets structure
+
+A property set consists of a reserved name, enclosed in square brackets,
+followed by a series of string key and value pairs. The set name and each entry
+in the set are separated by a newline.
+
+The string key and value pairs have the following format:
+```
+<string key>=<value type>|<value>
+```
+
+The value type is a string and the value of it has the following meaning for the
+corresponding value:
+
+| Value type | Description                               |
+| ---------- | ----------------------------------------- |
+| "0"        | The value has no known type.              |
+| "1"        | The value is a 32 bit integer.            |
+| "2"        | The value is a base64 encoded byte array. |
+
+### Property sets
+
+This section describes the known property sets.
+
+#### SYCL/specialization constants
+
+__Key:__ Specialization constant name.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ Information about the specialization constant with the following
+fields: 
+
+```c++
+// Encodes ID of a scalar specialization constants which is a leaf of some
+// composite specialization constant.
+unsigned ID;
+// Encodes offset from the beginning of composite, where scalar resides, i.e.
+// location of the scalar value within a byte-array containing the whole
+// composite specialization constant. If descriptor is used to represent a
+// whole scalar specialization constant instead of an element of a composite,
+// this field should be contain zero.
+unsigned Offset;
+// Encodes size of scalar specialization constant.
+unsigned Size;
+```
+
+
+#### SYCL/specialization constants default values
+
+__Key:__ Specialization constant name.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ Byte representation of the default value for the specialization
+constant.
+
+
+#### SYCL/devicelib req mask
+
+__Key:__ At most one entry with "DeviceLibReqMask".
+
+__Value type:__ 32 bit integer. ("1")
+
+__Value:__ A bitmask of which device libraries the binary uses.
+
+
+#### SYCL/kernel param opt
+
+__Key:__ Kernel name.
+
+__Value type:__ Byte array. ("2")
+
+__Value:__ A bitmask identifying the arguments of the kernel that have been
+removed by the dead-argument-elimination optimization pass.
+
+
+#### SYCL/program metadata
```

Comment:
This one seems underspecified.  Is this something you're adding as part of the SYCLBIN format, or is it an existing property set?

Reply:
No, this is old. It is used by CUDA and HIP to tell the runtime about things like max work-group size of kernels, i.e. `kernel_name@max_work_group_size` could be an entry.

Reply:
OK, not required to fix the specification in this PR, but this is technical debt that should be improved.  I think this specification will end up being the official description of all our property sets, so we should try to make it complete.